engine/grammar/matchers/datatypes/semver.go

```diff
@@ -0,0 +1,202 @@
+package datatypes
+
+import (
+	"fmt"
+	"strconv"
+
+	"strings"
+)
+
+const (
+	MetadataDelimiter   = "+"
+	PreReleaseDelimiter = "-"
+	ValueDelimiter      = "."
+)
+
+type Semver struct {
+	major     int64
+	minor     int64
+	patch     int64
+	preRealse []string
+	isStable  bool
+	metadata  string
+	version   string
+}
+
+func BuildSemver(version string) (*Semver, error) {
+	if len(strings.TrimSpace(version)) == 0 {
+		return nil, fmt.Errorf("unable to convert to semver, version cannot be empty")
+	}
+	metadata, vWithoutMetadata, err := processMetadata(version)
+	if err != nil {
+		return nil, err
+	}
+	preRealse, vWithoutPreRelease, err := processPreRelease(vWithoutMetadata)
+	if err != nil {
+		return nil, err
+	}
+	major, minor, patch, err := processComponents(vWithoutPreRelease)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Semver{
+		metadata:  metadata,
+		preRealse: preRealse,
+		major:     major,
+		minor:     minor,
+		patch:     patch,
+		isStable:  len(preRealse) == 0,
+		version:   version,
+	}, nil
+}
+
+func processMetadata(version string) (string, string, error) {
+	index, metadata := calculateIndexAndTrimString(version, MetadataDelimiter)
+	if index == -1 {
+		return "", version, nil
+	}
+
+	if len(metadata) == 0 {
+		return "", "", fmt.Errorf("unable to convert to semver, incorrect pre release data")
+
+	}
+	return metadata, strings.TrimSpace(version[0:index]), nil
+}
+
+func processPreRelease(vWithoutMetadata string) ([]string, string, error) {
+	index, preReleaseData := calculateIndexAndTrimString(vWithoutMetadata, PreReleaseDelimiter)
+	if index == -1 {
+		return nil, vWithoutMetadata, nil
+	}
+
+	preRealse := strings.Split(preReleaseData, ValueDelimiter)
+
+	if len(preRealse) == 0 || isEmpty(preRealse) {
+		return nil, "", fmt.Errorf("unable to convert to semver, incorrect pre release data")
+	}
+	return preRealse, strings.TrimSpace(vWithoutMetadata[0:index]), nil
+}
+
+func calculateIndexAndTrimString(vWithoutMetadata string, delimiter string) (int, string) {
```

Comment:
rename to `extract(str string, delimiter string) (int, string)`

Reply:
Updated.